src/raftstore/coprocessor/split_check/half.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use rocksdb::DB;
+
+use super::super::{Coprocessor, ObserverContext, SplitCheckObserver};
+use super::Status;
+
+const BUCKET_NUMBER_LIMIT: usize = 1024;
+#[derive(Default)]
+pub struct HalfStatus {
+    buckets: Vec<Vec<u8>>,
+    cur_bucket_size: u64,
+    each_bucket_size: u64,
+}
+
+impl HalfStatus {
+    fn new(each_bucket_size: u64) -> HalfStatus {
+        HalfStatus {
+            each_bucket_size: each_bucket_size,
+            ..Default::default()
+        }
+    }
+
+    pub fn push_data(&mut self, key: &[u8], value_size: u64) {
+        let current_len = key.len() as u64 + value_size;
+        if self.buckets.is_empty() || self.cur_bucket_size >= self.each_bucket_size {
+            self.check_and_adjust_buckets_num();
+            self.buckets.push(key.to_vec());
+            self.cur_bucket_size = 0;
+        }
+        self.cur_bucket_size += current_len;
+    }
+
+    fn check_and_adjust_buckets_num(&mut self) {
+        if self.buckets.len() < BUCKET_NUMBER_LIMIT {
+            return;
+        }
+        let buckets_num = self.buckets.len() / 2;
+        for id in { 0..buckets_num } {
+            self.buckets.swap(id, id * 2 + 1);
+        }
+        self.buckets.truncate(buckets_num);
+        self.each_bucket_size *= 2;
+    }
+
+    pub fn split_key(self) -> Option<Vec<u8>> {
+        let mid = self.buckets.len() / 2;
+        if mid == 0 {
```

Comment:
`buckets.get()` will take care of this.

Reply:
It's used to check `if self.buckets.len() <= 1` here